packages/common-common/src/permissions.ts

```diff
@@ -1,63 +1,428 @@
+      Action.DELETE_THREAD,
+      Action.LINK_PROPOSAL_TO_THREAD,
+    ],
+  ],
+  [
+    Action.LINK_PROPOSAL_TO_THREAD,
+    [
+      Action.CREATE_REACTION,
+      Action.VIEW_REACTIONS,
+      Action.DELETE_REACTION,
+      Action.CREATE_COMMENT,
+      Action.VIEW_COMMENTS,
+      Action.EDIT_COMMENT,
+      Action.DELETE_COMMENT,
+      Action.CREATE_POLL,
+      Action.VIEW_POLLS,
+      Action.VOTE_ON_POLLS,
+      Action.VIEW_THREADS,
+      Action.CREATE_THREAD,
+      Action.EDIT_THREAD,
+      Action.DELETE_THREAD,
+      Action.LINK_THREAD_TO_THREAD,
+    ],
+  ],
+  [
+    Action.VIEW_TOPICS,
+    [
+      Action.VIEW_REACTIONS,
+      Action.VIEW_COMMENTS,
+      Action.VIEW_POLLS,
+      Action.VIEW_THREADS,
+    ],
+  ],
+  [
+    Action.CREATE_TOPIC,
+    [
+      Action.CREATE_REACTION,
+      Action.VIEW_REACTIONS,
+      Action.DELETE_REACTION,
+      Action.CREATE_COMMENT,
+      Action.VIEW_COMMENTS,
+      Action.EDIT_COMMENT,
+      Action.DELETE_COMMENT,
+      Action.CREATE_POLL,
+      Action.VIEW_POLLS,
+      Action.VOTE_ON_POLLS,
+      Action.VIEW_THREADS,
+      Action.CREATE_THREAD,
+      Action.EDIT_THREAD,
+      Action.DELETE_THREAD,
+      Action.LINK_THREAD_TO_THREAD,
+      Action.LINK_PROPOSAL_TO_THREAD,
+      Action.VIEW_TOPICS,
+      Action.EDIT_TOPIC,
+      Action.DELETE_TOPIC,
+    ],
+  ],
+  [
+    Action.EDIT_TOPIC,
+    [
+      Action.CREATE_REACTION,
+      Action.VIEW_REACTIONS,
+      Action.DELETE_REACTION,
+      Action.CREATE_COMMENT,
+      Action.VIEW_COMMENTS,
+      Action.EDIT_COMMENT,
+      Action.DELETE_COMMENT,
+      Action.CREATE_POLL,
+      Action.VIEW_POLLS,
+      Action.VOTE_ON_POLLS,
+      Action.VIEW_THREADS,
+      Action.CREATE_THREAD,
+      Action.EDIT_THREAD,
+      Action.DELETE_THREAD,
+      Action.LINK_THREAD_TO_THREAD,
+      Action.LINK_PROPOSAL_TO_THREAD,
+      Action.VIEW_TOPICS,
+      Action.CREATE_TOPIC,
+      Action.DELETE_TOPIC,
+    ],
+  ],
+  [
+    Action.DELETE_TOPIC,
+    [
+      Action.CREATE_REACTION,
+      Action.VIEW_REACTIONS,
+      Action.DELETE_REACTION,
+      Action.CREATE_COMMENT,
+      Action.VIEW_COMMENTS,
+      Action.EDIT_COMMENT,
+      Action.DELETE_COMMENT,
+      Action.CREATE_POLL,
+      Action.VIEW_POLLS,
+      Action.VOTE_ON_POLLS,
+      Action.VIEW_THREADS,
+      Action.CREATE_THREAD,
+      Action.EDIT_THREAD,
+      Action.DELETE_THREAD,
+      Action.LINK_THREAD_TO_THREAD,
+      Action.LINK_PROPOSAL_TO_THREAD,
+      Action.VIEW_TOPICS,
+      Action.CREATE_TOPIC,
+      Action.EDIT_TOPIC,
+    ],
+  ],
+]);
 
-export enum PermissionError {
-  NOT_PERMITTED = 'Action not permitted',
+// Adds or Removes the implicit permissions of a permission
+export function addRemoveImplicitPermissions(
+  permission: Permissions,
+  actionNumber: number,
+  isAdd: boolean
+): Permissions {
+  let result = BigInt(permission);
+  const implicitActions = IMPLICIT_PERMISSIONS_BY_ACTION.get(actionNumber);
```

Comment:
This will need to become a recursive resolution given above change -- maybe add a helper function.